### ISSUE TITLE
Extra move added if number of moves = 7

### ISSUE THREAD

**Issue by MrDodiDEV** [MrDodiDEV/MLTraining#30]
`import torch
import torch.nn.functional as F
import cv2
import numpy as np
import matplotlib.pyplot as plt

GRID_WIDTH = 11
GRID_HEIGHT = 18
CELL_SIZE = 84
GRID_START_X = 150
GRID_START_Y = 15

def denormalize_coordinates(normalized_coordinates):
    x = torch.round(normalized_coordinates[0] * GRID_WIDTH)
    y = torch.round(normalized_coordinates[1] * GRID_HEIGHT)
    return torch.stack((x, y), dim=-1)

def extract_features_from_route(route_tensor, features_per_move):
    """Extract features from the route tensor, handling different feature dimensions."""
    features = []
    num_moves = route_tensor.shape[0] // features_per_move
    print(f"num_moves : {num_moves}")
    print(f"route_tensor.shape : {route_tensor.shape[0]}")

    for i in range(num_moves):
        start_idx = i * features_per_move
        end_idx = start_idx + features_per_move
        move_features = route_tensor[start_idx:end_idx]

        # Check for rows with more than 5 zeros
        if torch.sum(move_features == 0) <= 7:
            normalized_coordinates = move_features[:2]
            is_start = move_features[2]
            is_end = move_features[3]
            denormalized_coords = denormalize_coordinates(normalized_coordinates)
            features.append((denormalized_coords, is_start, is_end))
        else:
            print(f"Skipping move {i} due to excessive zeros.")
    return features

def visualize_routes(generated_features, original_features, image_path):
    img = cv2.imread(image_path)
    img_rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

    def apply_offsets(coords):
        x, y = coords
        pixel_x = GRID_START_X + int(x.item()) * CELL_SIZE
        pixel_y = img_rgb.shape[0] - (GRID_START_Y + int(y.item()) * CELL_SIZE)
        return pixel_x, pixel_y

    img_generated = img_rgb.copy()
    img_original = img_rgb.copy()

    def plot_route(features, image):
        for coords, is_start, is_end in features:
            pixel_x, pixel_y = apply_offsets(coords)
            is_start_val = is_start.item()
            is_end_val = is_end.item()

            color = (0, 151, 0) if is_start_val > 0.5 else (255, 0, 0) if is_end_val > 0.5 else (0, 0, 255)
            cv2.circle(image, (pixel_x, pixel_y), 45, color, 3)

    plot_route(generated_features, img_generated)
    plot_route(original_features, img_original)

    plt.figure(figsize=(20, 10))
    plt.subplot(1, 2, 1)
    plt.imshow(img_original)
    plt.title("Original Route")
    plt.axis('off')

    plt.subplot(1, 2, 2)
    plt.imshow(img_generated)
    plt.title("Generated Route")
    plt.axis('off')

    plt.tight_layout()
    plt.show()

    return img_original, img_generated

# Adjust feature dimensions per dataset
generated_features_per_move = 19
original_features_per_move = 20

generated_features = extract_features_from_route(generated_data.squeeze(), generated_features_per_move)
original_features = extract_features_from_route(existing_route.squeeze(), original_features_per_move)

img_original, img_generated = visualize_routes(generated_features, original_features, "/content/MoonBoardCropped_2017_SS.jpg")
`

Ha num_moves=7 akkor a route shape [0] az 152 ezért 8 moveot fog exractolni
